### DISCUSSION TITLE
Why are the serializers not constrained within the class/struct themselves?

### DISCUSSION THREAD

**Discussion by mr5z** [nlohmann/json#2849] [Q&A] [answered]
Just got back to developing with C++ again from higher level languages so I'm kinda hoping for stricter rules for custom types for the serializer/deserializer of this JSON library.

Why are the serializers not constrained within the class/struct itself?

What was the consensus why it is designed this way?

Probably related: #717

**Comment by nlohmann** [comment, marked as answer]
I don’t understand what you mean. Can you give an example?

**Reply by mr5z** [reply]
Ah sorry for not being clear. What I meant are these functions:
```cpp
void to_json(json& j, const T& s);
void from_json(const json& j, T& s);
```
I was thinking it could be member functions instead.

**Reply by nlohmann** [reply]
This is to allow extending existing types after the fact, e.g., containers from Boost or Qt without the need to access to the source code of the original classes.

**Reply by mr5z** [reply]
That make sense but is there any downside of supporting the OOP approach of this? Or do you see it just a boilerplate/unnecessary feature to implement?

Thanks for answering btw.

**Reply by nlohmann** [reply]
A member function approach would not be able to solve the issue with existing types, so it would only be syntactic sugar on top an already quite heavy approach.

**Reply by mr5z** [reply]
Hi again @nlohmann :)

I have encountered this compilation error where the body definition is "defined multiple times" because of how I did it.
So for every model, they are in a separate file within namespace, and the JSON serializer/deserializer is defined also within that namespace and file. A workaround would be adding the `inline` keyword with every non-member function according to this: https://github.com/nlohmann/json/issues/804.

So my question is, would you recommend this workaround? I was thinking we could avoid this if we just define the JSON serializer/deserializer as a member function.

**Reply by gregmarr** [reply]
They need to be free functions, so you do need to add the `inline` keyword for non-member functions if they are in a header file that is included in multiple compilation units.  This isn't a workaround, this is just how the language works.
